src/redux/filmList/sagas/filmList.saga.ts

```diff
@@ -27,9 +42,69 @@ export function* getFilmsTask(data: {
   payload: FilmOptions;
 }): Generator<StrictEffect, void, any> {
   try {
-    const options = data.payload;
+    let needClear = true;
+    let needReload = true;
+    const genre = yield select(genreSelector);
+    const searchTitle = yield select(searchedFilmSelector);
+    const offset = yield select(offsetSelector);
+    const sortingType = yield select(sortingTypeSelector);
+
+    const state = {
+      genre,
+      searchTitle,
+      offset,
+      sortingType,
+    };
+    const options = {
+      ...state,
+      ...data.payload,
+    };
+
+    switch (Object.keys(data.payload)[0]) {
+      case "searchTitle":
+        if (data.payload.searchTitle !== searchTitle) {
+          yield put(searchFilm(data.payload.searchTitle!));
+          options.offset = 0;
+        } else {
+          needReload = false;
+          needClear = false;
+        }
+
+        break;
+      case "genre":
+        if (data.payload.genre !== genre) {
+          yield put(changeGenre(data.payload.genre!));
+          options.offset = 0;
+        } else {
+          needReload = false;
+          needClear = false;
+        }
+
+        break;
+      case "sortingType":
+        if (data.payload.sortingType !== sortingType) {
+          yield put(changeGenre(data.payload.sortingType!));
+          options.offset = 0;
+        } else {
+          needReload = false;
+          needClear = false;
+        }
+        break;
+      case "offset":
+        yield put(getMoreFilms());
+        needClear = false;
+        options.offset = offset + FILM_LIMIT;
+        break;
+      default:
+        needClear = false;
+        break;
+    }
+
+    const res = needReload ? yield call(getFilms, options) : [];
+    if (needClear) {
+      yield put(clearfilmsList());
+    }
```

Comment:
i've meant different approach, on these actions (searchFilm, changeGenre,changeSorting ..) you can have the same task handler 
```
export function* filmsRootSaga() {
  yield takeLatest(searchFilm, getFilmsTask);
  yield takeLatest(changeGenre, getFilmsTask);
 ....
}
```
And then each action, e.g, `searchFilm` can set all options in theirs payload which you are setting in switch right now. So in your component, you can do like,  
```
const onSearchFilm = useCallback((event) => {
    event.preventDefault();
    const input = event.target[0].value;

    dispatch(searchFilm(fetchOptions: { searchTitle: input, offset: someValue, etc,}, shouldReload: false, shouldClear: false}));
  }, []);
```
and then your task would be much more smaller.

```
export function* getFilmsTask(data: {
  payload: {
  options: FilmOptions,
  shouldReload: boolean,
  shouldClear: boolean
} ;
}): Generator<StrictEffect, void, any> {
  try {
  
    const {payloadOptions,  shouldReload = true, shouldClear = true} = data.payload;
 
    // make 1 selector which gets all options from redux as one object;
   const stateOptions = yield select(optionsSelector);

    const options = {
      ...stateOptions ,
      ...payloadOptions,
    };
    yield put(fetchfilmsList.request)
    const res = shouldReload? yield call(getFilms, options) : [];
    if (shouldClear) {
      yield put(clearfilmsList());
    }
...
```